README.md
# chlorpromazineR: Convert antipsychotic doses to chlorpromazine equivalents

[![cran checks](https://cranchecks.info/badges/summary/chlorpromazineR)](https://cran.r-project.org/web/checks/check_results_chlorpromazineR.html) [![R-CMD-check](https://github.com/ropensci/chlorpromazineR/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/chlorpromazineR/actions) [![Coverage status](https://codecov.io/gh/ropensci/chlorpromazineR/branch/master/graph/badge.svg)](https://codecov.io/github/ropensci/chlorpromazineR?branch=master) [![DOI](https://zenodo.org/badge/175675220.svg)](https://zenodo.org/badge/latestdoi/175675220) [![](https://badges.ropensci.org/307_status.svg)](https://github.com/ropensci/software-review/issues/307/) 


Studies investigating or controlling for the impact of antipsychotic medications often need to quantify the amount of medication to which an individual is or has been exposed. As different antipsychotics have different potencies, the task is more complicated than using each medication’s daily dosage in milligrams, for example. `chlorpromazineR` is an R package to calculate dose equivalents for common oral and injectable antipsychotic medications based on conversion factors from the published literature. We do not propose to suggest which conversion factors are appropriate to use, or how to interpret the converted data. All users should also refer to the papers from which the conversion factor data originates to determine whether the use of such data is appropriate for their study.

We hope that this package is of use to scientists who do clinical research involving antipsychotic medications. Specifically, the goals of this package are:

* to improve transparency and consistency in calculating chlorpromazine equivalents,
* to reduce human error and improve accuracy,
* and to simplify workflows for large datasets, as from chart reviews of electronic health records.

For further details and usage, please see the [walkthrough vignette](https://htmlpreview.github.io/?https://github.com/ropensci/chlorpromazineR/blob/master/doc/walkthrough.html).

This results from this package should be double checked for accuracy when used in production. We welcome feedback--please contact via eb@ericebrown.com or file an issue.

## Installation

The CRAN release version (recommended) can be installed via the command: `install.packages("chlorpromazineR")`.

The development version of this package can be installed via the command: `devtools::install_github("ropensci/chlorpromazineR")`.

## Usage

Once installed, load package with `library(chlorpromazineR)`. The package's main conversion functions are documented and usage and examples can be seen with `help(to_cpz)` and `help(to_ap)`. It is strongly recommended to read the articles from which the keys are derived, and to verify that the program's output is producing results as expected. This package facilitates bulk conversion, but should be verified to ensure it produces the results as expected.

### Online calculator

An online calculator using this package is available as a [shiny app](http://ap.eebc.ca/).

### Convert data to chlorpromazine equivalents

    participant_ID <- c("P01", "P02", "P03", "P04")
    age <- c(42, 29, 30, 60) # not used in calculation
    antipsychotic <- c("olanzapine", "olanzapine", "quetiapine", "ziprasidone")
    dose <- c(10, 12.5, 300, 60)
    example_oral <- data.frame(participant_ID, age, antipsychotic, dose, 
                               stringsAsFactors = FALSE)
    to_cpz(example_oral, ap_label = "antipsychotic", dose_label = "dose", 
           route = "oral")

## Disclaimer

This package is not for clinical use. The authors assume no liability. All work must be verified independently. Use at own risk.

## Licence

Copyright (C) 2019-2021 Eric E. Brown. This software is licensed under the GPL-3.

## Citation

If you use this package in your scientific paper, please cite this package and the original papers from which the conversion factors are derived. The references can be viewed by using the built-in help function, e.g. `help(gardner2010)`, and as listed below. In addition, a spreadsheet-based tool facilitating dose equivalence conversion has been published by Leucht et al. (2020).

    Davis, J. (1974). Dose equivalence of the anti-psychotic drugs.
    Journal of Psychiatric Research, 11, 65-69.
    <https://doi.org/10.1016/0022-3956(74)90071-5>

    Gardner, D. M., Murphy, A. L., O’Donnell, H., Centorrino, F., & 
    Baldessarini, R. J. (2010). International consensus study of 
    antipsychotic dosing. The American Journal of Psychiatry, 167(6),
    686–693. <https://doi.org/10.1176/appi.ajp.2009.09060802>

    Leucht, S., Samara, M., Heres, S., & Davis, J. M. (2016). Dose
    Equivalents for Antipsychotic Drugs: The DDD Method. Schizophrenia
    Bulletin, 42(suppl_1), S90–S94. <https://doi.org/10.1093/schbul/sbv167>
    
    Leucht, S., Crippa, A., Siafis, S., Patel, M., Orsini, N. & Davis, J. M. 
    (2020). Dose-Response Meta-Analysis of Antipsychotic Drugs for Acute 
    Schizophrenia. American Journal of Psychiatry. 117(4).
    <https://doi.org/10.1176/appi.ajp.2019.19010034>

    Woods, S. (2003). Chlorpromazine Equivalent Doses for the Newer
    Atypical Antipsychotics. Journal of Clinical Psychiatry. 64(6).
    663-667. <https://doi.org/10.4088/JCP.v64n0607>
    
    

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
# chlorpromazineR 0.1.2

This is the first CRAN release of the package. The package has just passed peer-review through rOpenSci. It is published in a fully functional state.

# chlorpromazineR 0.1.3

## Major changes

Added new vignette with examples from each of the keys (references).

## Minor changes

Added additional tests and code cleanup.

# chlorpromazineR 0.2.0

## Major changes

Added new key: leucht2020 (<https://doi.org/10.1176/appi.ajp.2019.19010034>)

### Minor changes

Added additional tests to compare results to spreadsheet by Leucht et al. with similar purpose.


tests/testthat.R
library(testthat)
library(chlorpromazineR)

test_check("chlorpromazineR")


tests/testthat/test_key.R
# chlorpromazineR package
# See README.md
# test_key.R

context("checking key manipulation functions")

test_that("trim_key() works with gardner2010 data", {

  trimmed <- trim_key(gardner2010_withsai)

  original_sai <-  c("chlorpromazine hcl", "clotiapine injectable", 
                     "fluphenazine hcl", "haloperidol lactate", "loxapine hcl")

  expected_sai <- c("chlorpromazine", "clotiapine", "fluphenazine", 
                    "haloperidol", "loxapine")

  original_lai <- c("clopenthixol decanoate", "flupenthixol decanoate",  
                    "fluphenazine decanoate", "fluphenazine enanthate", 
                    "fluspirilene")

  expected_lai <- c("clopenthixol", "flupenthixol", "fluphenazine", 
                    "fluphenazine", "fluspirilene")

  expect_equal(names(gardner2010_withsai$sai)[1:5], original_sai)
  expect_equal(names(trimmed$sai)[1:5], expected_sai)
  expect_equal(names(gardner2010_withsai$lai)[1:5], original_lai)
  expect_equal(names(trimmed$lai)[1:5], expected_lai)
})

test_that("add_key() works with gardner2010 and leucht2016", {

  merged <- add_key(gardner2010, leucht2016, trim=TRUE)

  expect_equal(merged$oral$haloperidol, gardner2010$oral$haloperidol)
  expect_equal(merged$oral$tiapride, leucht2016$oral$tiapride)

  expect_warning(add_key(gardner2010, leucht2016, trim=FALSE))

})

test_that("has_long_name() does as it says", {

  expect_equal(has_long_name(gardner2010), TRUE)
  expect_equal(has_long_name(gardner2010_withsai), TRUE)
  expect_equal(has_long_name(leucht2016), FALSE)
    
})

test_that("included keys are valid format", {
  expect_equal(check_key(gardner2010), TRUE)
  expect_equal(check_key(gardner2010_withsai), TRUE)
  expect_equal(check_key(leucht2016), TRUE)
  expect_equal(check_key(davis1974), TRUE)
  expect_equal(check_key(woods2003), TRUE)
})


tests/testthat/test_checks.R
# chlorpromazineR package
# See README.md
# test_checks.R

context("checking functions behave")

test_that("check_ap() works with all matching example", {

  antipsychotic <- c("Amisulpride", "Aripiprazole", "Benperidol", 
                     "Chlorpromazine", "Clopenthixol",

                     "Chlorpromazine HCl", "Clotiapine injectable", 
                     "Fluphenazine HCl", "Haloperidol lactate",
                     "Loxapine HCl",

                     "Clopenthixol decanoate", "Flupenthixol decanoate", 
                     "Fluphenazine decanoate", "Fluphenazine enanthate", 
                     "Fluspirilene")

  dose <- c(700, 30, 5, 600, 60,
            100, 40, 5, 5, 25,
            300, 40, 25, 25, 6)

  route <- c(rep("oral", 5), rep("sai", 5), rep("lai", 5))

  q <- c(14,14,14,14,7)

  example <- data.frame(antipsychotic, dose, q, route,
                        stringsAsFactors = FALSE)

  result <- check_ap(example, gardner2010_withsai, "antipsychotic", "mixed",
                     "route")

  expect_equal(0, result)

})

test_that("check_ap() works with single route example, sai", {

  antipsychotic <- c("Chlorpromazine HCl", "Clotiapine injectable", 
                     "Fluphenazine HCl", "Haloperidol lactate",
                     "Loxapine HCl")

  dose <- c(100, 40, 5, 5, 25)

  example <- data.frame(antipsychotic, dose, stringsAsFactors = FALSE)

  result <- check_ap(example, gardner2010_withsai, "antipsychotic", "sai")

  expect_equal(0, result)

  antipsychotic[2] <- "not real"
  example <- data.frame(antipsychotic, dose, stringsAsFactors = FALSE)
  result <- check_ap(example, gardner2010_withsai, "antipsychotic", "sai")
  expect_equal(1, result)


})

test_that("check_ap() works with single route example, oral", {
    
    antipsychotic <- c("Amisulpride", "Aripiprazole", "Benperidol",
                       "Chlorpromazine", "Clopenthixol")
    
    dose <- c(700, 30, 5, 600, 60)
    
    example <- data.frame(antipsychotic, dose, stringsAsFactors = FALSE)
    
    result <- check_ap(example, gardner2010, "antipsychotic", "oral")
    
    expect_equal(0, result)
    
    antipsychotic[2] <- "not real"
    example <- data.frame(antipsychotic, dose, stringsAsFactors = FALSE)
    result <- check_ap(example, gardner2010, "antipsychotic", "oral")
    expect_equal(1, result)
    
    
})

test_that("check_ap() works with mismatches in each route", {

  antipsychotic <- c("Amisulpride", "Aripiprazole", "Benperidol", 
                     "Chlorpromazine HCl", "Clopenthixol",

                     "Chlorpromazine HCl", "Clotiapine injectable", 
                     "Fluphenazine HCl", "Haloperidol",
                     "Loxapine HCl",

                     "Clopenthixol decanoate", "Flupenthixol", 
                     "Fluphenazine decanoate", "Fluphenazine enanthate", 
                     "Fluspirilene")

  dose <- c(700, 30, 5, 600, 60,
            100, 40, 5, 5, 25,
            300, 40, 25, 25, 6)

  route <- c(rep("oral", 5), rep("sai", 5), rep("lai", 5))

  q <- c(14,14,14,14,7)

  example <- data.frame(antipsychotic, dose, q, route,
                        stringsAsFactors = FALSE)

  result <- check_ap(example, gardner2010_withsai, "antipsychotic", "mixed",
                     "route")

  expect_equal(3, result)

})

test_that("included keys validate with check_key()", {

  expect_equal(TRUE, check_key(gardner2010))
  expect_equal(TRUE, check_key(leucht2016))

})

test_that("bad keys don't validate with check_key()", {

  expect_equal(TRUE, check_key(gardner2010))
  expect_equal(TRUE, check_key(leucht2016))

  gbad <- gardner2010
  names(gbad) <- c("oral", "sai", "depot")
  expect_error(check_key(gbad))

  expect_error(check_key(list(oral=1, sai=2)))

  expect_error(check_key(4))
  expect_error(check_key(c(4,4,4)))

  close <- list(oral=1,sai=2,lai=3)
  expect_error(check_key(close))

  closer <- list(oral=list(a="hi"),sai=list(a="hi"),lai=list(a="hi"))
  expect_error(check_key(closer))
})

test_that("trim_key() doesn't try to work on bad key", {

  expect_error(trim_key(3))

})

test_that("add_key() doesn't try to work on bad key", {

  expect_error(add_key(gardner2010, 3, trim = T))
  expect_error(add_key(3, gardner2010, trim = T))

})



R/convert.R
# chlorpromazineR package
# See README.md

#' Calculates chlorpromazine-equivalent doses
#' 
#' Given a data.frame containing doses of antipsychotics to_cpz() converts the
#' doses into the equivalent chlorpromazine (CPZ) doses, using the conversion
#' factors specified in the key.
#'
#' The default key is gardner2010 which has data for both oral and long-acting
#' antipsychotic medications. See help(gardner2010) for the source reference.
#' 
#' @export
#' @param input_data data.frame with antipsychotic name and dose data
#' @param ap_label column in x that stores antipsychotic name
#' @param dose_label column in x that stores dose
#' @param route options include "oral", "sai", "lai" or "mixed"
#' @param key source of the conversion factors--defaults to Gardner et al. 2010
#' @param eq_label the name of the column to be created, to save the 
#' calculated CPZ-equivalent dose 
#' @param factor_label the name of the column to be created to store the
#' conversion factors
#' @param route_label if "mixed" route is specified, provide the column that
#' stores the route information
#' @param q_label if long-acting injectable doses are included, provide the 
#' column that stores the injection frequency (days), or only if the doses have
#' already been divided, set q_label = 1.
#' @return data.frame with new variables storing conversion factor and 
#' CPZ-equivalent doses
#' @family conversion functions
#' @examples
#' participant_ID <- c("P01", "P02", "P03", "P04")
#' age <- c(42, 29, 30, 60)
#' antipsychotic <- c("olanzapine", "olanzapine", "quetiapine", "ziprasidone")
#' dose <- c(10, 12.5, 300, 60)
#' example_oral <- data.frame(participant_ID, age, antipsychotic, dose, 
#'                            stringsAsFactors = FALSE)
#' to_cpz(example_oral, ap_label = "antipsychotic", dose_label = "dose", 
#'        route = "oral")
to_cpz <- function(input_data, ap_label, dose_label, route="oral",
                   key=chlorpromazineR::gardner2010, eq_label="cpz_eq", 
                   factor_label="cpz_conv_factor", route_label=NULL,
                   q_label=NULL) {
  
  x <- input_data
  
  check_params(x, ap_label, dose_label, route, eq_label, factor_label,
               route_label, q_label)
  check_key(key)
  
  if (check_ap(x, key = key, ap_label = ap_label, route = route,
               route_label = route_label) != 0) {
    stop("Data contains antipsychotics not in the key")
  }
  
  cpz_conv_factor <- data.frame(NA)
  cpz_eq <- data.frame(NA)
  names(cpz_conv_factor) <- factor_label
  names(cpz_eq) <- eq_label
  x <- cbind(x, cpz_conv_factor, cpz_eq)
    
  x[,ap_label] <- tolower(x[,ap_label])
  
  if (route %in% c("oral", "sai", "lai")) {
    x[, factor_label] <- as.numeric(key[[route]][x[, ap_label]])
    x[, eq_label] <- apply(x[, c(dose_label, factor_label)], 1, prod)

    if (route == "lai") {
      if (is.null(q_label)) {
        stop("q_label must be specified for lai route.")
      }
      if (q_label != 1) {
        x[, eq_label] <- x[, eq_label] / x[, q_label]
    }
  } 
                                                                   
  }

  if (route == "mixed") {
        
    if (!check_route(x, route_label)) stop("Route column must only include
                                            oral, sai or lai")
        
    if (is.null(q_label)) stop("A column name for the LAI frequency, q_label,
                                (days) must be specified")
    if (is.null(route_label)) stop("A column name for the route, 
                                   route_label, must be specified.")
    if (!is.numeric(x[, q_label])) stop("q_label column must be numeric for
                                         LAIs (days)")
        
    x <- convert_by_route(x=x, key=key, ap_label=ap_label, 
                          dose_label=dose_label, route_label=route_label, 
                          factor_label=factor_label, eq_label=eq_label)

    if (q_label != 1) {
      x[x$route == "lai", ][, eq_label] <-
                                  x[x[, route_label] == "lai", ][, eq_label] /
                                  x[x[, route_label] == "lai", ][, q_label]
      }
  }
    
  return(x)
}

#' @noRd
convert_by_route <- function(x, key, ap_label, dose_label, route_label, 
                             factor_label, eq_label) {

  for (r in c("oral", "sai", "lai")) {
    k <- key[[r]]
    x[x[, route_label] == r, ][, factor_label] <-
                           as.numeric(k[x[x[, route_label] == r,][, ap_label]])
    
    x[x[,route_label] == r, ][, eq_label] <-
         apply(x[x[, route_label] == r,][,c(dose_label, factor_label)], 1, prod)

  }
  
  return(x)
}


#' Checks whether antipsychotic names are in the key
#' 
#' Provided a data.frame, x, this checks that the antipsychotic names stored in
#' the x's variable ap_label are present in the key.
#' 
#' @export
#' @param input_data data.frame with antipsychotic name and dose data
#' @param key source of the conversion factors--defaults to Gardner et al. 2010
#' @param ap_label column in x that stores antipsychotic name
#' @param route options include "oral", "sai", "lai" or "mixed"
#' @param route_label if "mixed" route is specified, provide the column that
#' stores the route information
#' @return number of antipsychotic names in x[,ap_label] that don't match key
#' @family checking functions
#' @examples
#' participant_ID <- c("P01", "P02", "P03", "P04")
#' age <- c(42, 29, 30, 60) # not used in calculation, just shows other data
#'                          # can exist in the data.frame
#' antipsychotic <- c("olanzapine", "olanzapine", "quetiapine", "ziprasidone")
#' dose <- c(10, 12.5, 300, 60)
#' example_oral <- data.frame(participant_ID, age, antipsychotic, dose, 
#'                            stringsAsFactors = FALSE)
#' check_ap(example_oral, ap_label = "antipsychotic", route = "oral", 
#'          key = gardner2010)
check_ap <- function(input_data, key=chlorpromazineR::gardner2010, ap_label,
                     route, route_label) {
  
  x <- input_data
  check_key(key)
  
  if (route %in% c("oral", "sai", "lai")) {
      notfound <- !(tolower(x[, ap_label]) %in% names(key[[route]]))
      bad <- paste0(x[, ap_label][notfound], " (", route, ")\n")
  } else if (route == "mixed") {
      notfound_oral <- !(tolower(x[x[, route_label] == "oral",][, ap_label])
                         %in% names(key$oral))
      notfound_sai <- !(tolower(x[x[, route_label] == "sai",][, ap_label])
                        %in% names(key$sai))
      notfound_lai <- !(tolower(x[x[, route_label] == "lai",][, ap_label])
                        %in% names(key$lai))
      notfound <- c(notfound_oral, notfound_sai, notfound_lai)
        
      if (any(notfound_oral)) {
          bad1 <-
                paste(x[x[, route_label] == "oral",][, ap_label][notfound_oral],
                "(oral)\n")
      } else bad1 <- NULL

      if (any(notfound_sai)) {
          bad2 <- paste(x[x[, route_label] == "sai",][, ap_label][notfound_sai],
                        "(sai)\n")
      } else bad2 <- NULL

      if (any(notfound_lai)) {
          bad3 <- paste(x[x[, route_label] == "lai",][, ap_label][notfound_lai],
                        "(lai)\n")
      } else bad3 <- NULL

      bad <- c(bad1, bad2, bad3)
  } else {
    stop("route parameter must be oral, sai or lai.")
  }
    
  if (sum(notfound) > 0) {
      message("The following antipsychotics were not found in the key:")
      message(bad)
  }
    
  return(sum(notfound))
}

#' Calculates equivalent doses
#'
#' As in to_cpz(), to_ap() converts doses of antipsychotics into equivalent
#' doses to a reference antipsychotic. Whereas in to_cpz() the reference
#' antipsychotic is chlorpromazine (CPZ), to_ap() converts to equivalents of an
#' arbitrary antipsychotic specified as a string to convert_to_ap. Conversion
#' factors are specified in the key.
#'
#' @export
#' @param input_data data.frame with antipsychotic name and dose data
#' @param convert_to_ap name of desired reference antipsychotic
#' @param convert_to_route the route of the desired reference antipsychotic
#' @param ap_label column in x that stores antipsychotic name
#' @param dose_label column in x that stores dose
#' @param route options include "oral", "sai", "lai" or "mixed"
#' @param key source of the conversion factors--defaults to Gardner et al. 2010
#' @param cpz_eq_label the name of the column to be created, to save the
#' calculated CPZ-equivalent dose
#' @param ref_eq_label the name of the column to be created to save the doses in
#' terms of the specified reference antipsychotic (in convert_to_ap)
#' @param factor_label the name of the column to be created to store the
#' conversion factors
#' @param route_label if "mixed" route is specified, provide the column that
#' stores the route information
#' @param q_label if long-acting injectable doses are included, provide the
#' column that stores the injection frequency (days), or only if the doses have
#' already been divided, set q_label = 1.
#' @return data.frame with new variables storing conversion factor and
#' CPZ-equivalent doses
#' @family conversion functions
#' @examples
#' participant_ID <- c("P01", "P02", "P03", "P04")
#' age <- c(42, 29, 30, 60) # not used in calculation, just shows other data
#'                          # can exist in the data.frame
#' antipsychotic <- c("olanzapine", "olanzapine", "quetiapine", "ziprasidone")
#' dose <- c(10, 12.5, 300, 60)
#' example_oral <- data.frame(participant_ID, age, antipsychotic, dose,
#'                            stringsAsFactors = FALSE)
#' to_ap(example_oral, convert_to_ap="olanzapine", convert_to_route="oral",
#'       ap_label = "antipsychotic", dose_label = "dose", route = "oral")
to_ap <- function(input_data, convert_to_ap="olanzapine",
                  convert_to_route="oral", ap_label, dose_label, route="oral",
                  key=chlorpromazineR::gardner2010, cpz_eq_label="cpz_eq",
                  ref_eq_label="ap_eq", factor_label="cpz_conv_factor",
                  route_label=NULL, q_label=NULL) {
  
  if (route == "sai") {
    if (convert_to_route != "sai") {
      stop("For \"sai\" route, convert_to_route must be \"sai\" as well.")
    }
  }

  check_params(input_data, ap_label, dose_label, route, eq_label=cpz_eq_label,
               factor_label, route_label, q_label)
  check_key(key)
  
  if (!(convert_to_ap %in% names(key[[convert_to_route]]))) {
    stop("The specified convert_to antipsychotic/route is not in the key")
  }

  if (route == "sai") {
    if (convert_to_route != "sai") {
      stop("For \"sai\" route, convert_to_route must be \"sai\" as well.")
    }
  }
  
  out <- to_cpz(input_data = input_data, ap_label = ap_label,
                dose_label = dose_label, route = route, key = key,
                eq_label = cpz_eq_label, factor_label = factor_label,
                route_label = route_label, q_label = q_label)
                
  out[, ref_eq_label] <- out[, cpz_eq_label] /
                              as.numeric(key[[convert_to_route]][convert_to_ap])
  
  return(out)
}

#' @noRd
check_params <- function(x, ap_label, dose_label, route, eq_label,
                         factor_label, route_label, q_label) {
  if (!(is.data.frame(x))) stop("x must be a data.frame.")
  
  if (!(ap_label %in% names(x))) stop("ap_label must be a variable in x.")
  
  if (!(dose_label %in% names(x))) stop("dose_label must be a variable in x.")
  
  if ((length(route) != 1) | !(route %in% c("oral", "sai", "lai", "mixed"))) {
    stop("route must be 1 of \"oral\", \"sai\", \"lai\", or \"mixed\"")
  }
  
  if (route == "mixed") {
    if (!(route_label %in% names(x))) {
      stop("route_label must be a variable in x.")
    }
  } else {
    if (!(is.null(route_label))) {
      stop("route_label should be null if route is not mixed.")
    }
  }

  if (!(is.character(eq_label))) {
    stop("eq_label/cpz_eq_label must be a character string.")
  }
  
  if (!(is.character(factor_label))) {
    stop("factor_label must be a character string.")
  }
}

#' @noRd
check_route <- function(x, route_label) {
  return(all(x[,route_label] %in% c("oral", "sai", "lai")))
}


R/key_data.R
# key_data.R

# KEYS

#' Chlorpromazine equivalent key from Gardner et al. 2010 data
#'
#' A list of antipsychotics and their chlorpromazine equivalent doses, generated
#' from the following file included with the package:
#' system.file("extdata", "gardner2010.csv", package="chlorpromazineR").
#'
#' The SAI data is not included in this key, because the original study did not
#' specify a conversion factor from chlorpromazine LAI to oral. The alternative
#' key gardner2010_withsai can be used, which includes the SAI data, but the
#' chlorpromazine equivalent doses produced are equivalent to chlorpromazine SAI
#' not chlorpromazine oral. They could be manually converted (e.g. by
#' multiplying the SAI doses by 3 per equivalence noted by Davis 1974
#' <https://doi.org/10.1016/0022-3956(74)90071-5>)
#'
#' @format A named list of 3 named lists (1 for each route) and each sub-list
#' contains the conversion factors for each antipsychotic. The 3 top-level lists
#' are named `oral`, `sai`, and `lai` (route), and the lists they contain have
#' names corresponding to the antipsychotic, e.g. `olanzapine`.
#' @source Gardner, D. M., Murphy, A. L., O’Donnell, H., Centorrino, F., &
#' Baldessarini, R. J. (2010). International consensus study of antipsychotic
#' dosing. The American Journal of Psychiatry, 167(6), 686–693.
#' <https://doi.org/10.1176/appi.ajp.2009.09060802>
"gardner2010"

#' Chlorpromazine equivalent key from Gardner et al. 2010 data
#'
#' A list of antipsychotics and their chlorpromazine equivalent doses, generated
#' from the following file included with the package:
#' system.file("extdata", "gardner2010.csv", package="chlorpromazineR").
#'
#' The SAI equivalents produced by this key are equivalent to chlorpromazine SAI
#' not oral. They could be manually converted.
#'
#' @format A named list of 3 named lists (1 for each route) and each sub-list
#' contains the conversion factors for each antipsychotic. The 3 top-level lists
#' are named `oral`, `sai`, and `lai` (route), and the lists they contain have
#' names corresponding to the antipsychotic, e.g. `olanzapine`.
#' @source Gardner, D. M., Murphy, A. L., O’Donnell, H., Centorrino, F., &
#' Baldessarini, R. J. (2010). International consensus study of antipsychotic
#' dosing. The American Journal of #' Psychiatry, 167(6), 686–693.
#' <https://doi.org/10.1176/appi.ajp.2009.09060802>
"gardner2010_withsai"

#' Chlorpromazine equivalent key from Leucht et al. 2016 data
#'
#' A list of antipsychotics and their chlorpromazine equivalent doses, generated
#' from the following file included with the package:
#' system.file("extdata", "leucht2016.csv", package="chlorpromazineR").
#'
#' @format A named list of 3 named lists (1 for each route) and each sub-list
#' contains the conversion factors for each antipsychotic. The 3 top-level lists
#' are named `oral`, `sai`, and `lai` (route), and the lists they contain have
#' names corresponding to the antipsychotic, e.g. `olanzapine`.
#' @source Leucht, S., Samara, M., Heres, S., & Davis, J. M. (2016). Dose
#' Equivalents for Antipsychotic Drugs: The DDD Method. Schizophrenia Bulletin,
#' 42(suppl_1), S90–S94. <https://doi.org/10.1093/schbul/sbv167>
"leucht2016"

#' Antipsychotic equivalent key from Leucht et al. 2020 data
#'
#' A list of antipsychotics and their olanzapine-equivalent doses, generated
#' from the following file included with the package:
#' system.file("extdata", "leucht2020.csv", package="chlorpromazineR").
#'
#' This reference does not include chlorpromazine, so the conversion from
#' leucht2016 is implied (i.e. chlorpromazine = olanzapine * 30).
#'
#' @format A named list of 3 named lists (1 for each route) and each sub-list
#' contains the conversion factors for each antipsychotic. The 3 top-level lists
#' are named `oral`, `sai`, and `lai` (route), and the lists they contain have
#' names corresponding to the antipsychotic, e.g. `olanzapine`.
#' @source Leucht, S., Crippa, A., Siafis, S., Patel, M., Orsini, N. & Davis,
#' J. M. (2020). Dose-Response Meta-Analysis of Antipsychotic Drugs for Acute
#' Schizophrenia. American Journal of Psychiatry. 117(4).
#' <https://doi.org/10.1176/appi.ajp.2019.19010034>
"leucht2020"

#' Chlorpromazine equivalent key from Davis 1974 data
#'
#' A list of antipsychotics and their chlorpromazine equivalent doses, generated
#' from the following file included with the package:
#' system.file("extdata", "davis1974.csv", package="chlorpromazineR").
#'
#' @format A named list of 3 named lists (1 for each route) and each sub-list
#' contains the conversion factors for each antipsychotic. The 3 top-level lists
#' are named `oral`, `sai`, and `lai` (route), and the lists they contain have
#' names corresponding to the antipsychotic, e.g. `olanzapine`.
#' @source John Davis (1974). Dose equivalence of the anti-psychotic drugs.
#' Journal of Psychiatric Research, 11, 65-69.
#' <https://doi.org/10.1016/0022-3956(74)90071-5>
"davis1974"

#' Chlorpromazine equivalent key from Woods 2003 data
#'
#' A list of antipsychotics and their chlorpromazine equivalent doses, generated
#' from the following file included with the package:
#' system.file("extdata", "woods2003.csv", package="chlorpromazineR").
#'
#' @format A named list of 3 named lists (1 for each route) and each sub-list
#' contains the conversion factors for each antipsychotic. The 3 top-level lists
#' are named `oral`, `sai`, and `lai` (route), and the lists they contain have
#' names corresponding to the antipsychotic, e.g. `olanzapine`.
#' @source Scott Woods (2003). Chlorpromazine Equivalent Doses for the Newer
#' Atypical Antipsychotics. Journal of Clinical Psychiatry. 64(6). 663-667.
#' <https://doi.org/10.4088/JCP.v64n0607>
"woods2003"

# gardner2010gen <- function() {
#   g <- read.csv(system.file("extdata", "gardner2010.csv",
#                             package="chlorpromazineR"),
#                 stringsAsFactors = FALSE)
#   g_oral <- as.list(600 / g[g[,"route"]=="oral",]$median)
#   names(g_oral) <- tolower(g[g[,"route"]=="oral",]$drug)
#   g_sai <- as.list(100 / g[g[,"route"]=="SAI",]$median)
#   names(g_sai) <- tolower(g[g[,"route"]=="SAI",]$drug)
#   g_lai <- as.list((600 / ((g[g[,"route"]=="LAI",]$median) /
#                                      g[g[,"route"]=="LAI",]$days)))
#   names(g_lai) <- tolower(g[g[,"route"]=="LAI",]$drug)
#   return(list(oral=g_oral, sai=g_sai, lai=g_lai))
# }

# leucht2016gen <- function() {
# l <- read.csv(system.file("extdata", "leucht2016.csv",
#                           package="chlorpromazineR"),
#               stringsAsFactors = FALSE)
# l_oral <- as.list(100 / l[l[,"route"]=="oral",]$CPZ100MGEQ)
# names(l_oral) <- tolower(l[l[,"route"]=="oral",]$drug)
# l_sai <- as.list(100 / l[l[,"route"]=="SAI",]$CPZ100MGEQ)
# names(l_sai) <- tolower(l[l[,"route"]=="SAI",]$drug)
# l_lai <- as.list(100 / l[l[,"route"]=="LAI",]$CPZ100MGEQ)
# names(l_lai) <- tolower(l[l[,"route"]=="LAI",]$drug)
# return(list(oral=l_oral, sai=l_sai, lai=l_lai))
#}

#davis1974gen <- function() {
#  d <- read.csv(system.file("extdata", "davis1974.csv",
#                            package="chlorpromazineR"),
#                stringsAsFactors = FALSE)
#  d_oral <- as.list(100 / d$CPZ100)
#  names(d_oral) <- tolower(d$Antipsychotic)
#  d_sai <- as.list(300 / d$CPZ100)
#  names(d_sai) <- names(d_oral)
#  d_lai <- list()
#  return(list(oral=d_oral, sai=d_sai, lai=d_lai))
#}

#woods2003gen <- function() {
#  w <- read.csv(system.file("extdata", "woods2003.csv",
#                package="chlorpromazineR"),
#                stringsAsFactors = FALSE)
#  w_oral <- as.list(100 / w$CPZ100)
#  names(w_oral) <- tolower(w$Antipsychotic)
#  w_sai <- list()
#  w_lai <- list()
#  return(list(oral=w_oral, sai=w_sai, lai=w_lai))
#}

# leucht2020gen <- function() {
#   g <- read.csv(system.file("extdata", "leucht2020.csv",
#                             package="chlorpromazineR"),
#                 sep=",", header=TRUE)
#
#   g_oral <- as.list((1 / g[g[,"route"]=="oral",]$OLZ1mgeq) * 30)
#   names(g_oral) <- tolower(g[g[,"route"]=="oral",]$drug)
#
#   g_lai <- as.list((1 / ((g[g[,"route"]=="LAI",]$OLZ1mgeq))) * 30)
#   names(g_lai) <- tolower(g[g[,"route"]=="LAI",]$drug)
#
#   return(list(oral=g_oral, sai=list(), lai=g_lai))
# }
